pkg/tls/reloader.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2017 Frederic Branczyk All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tls
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"hash/fnv"
+	"io/ioutil"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+// CertReloader is the struct that parses a certificate/key pair,
+// providing a goroutine safe GetCertificate method to retrieve the parsed content.
+//
+// The GetCertificate signature is compatible with https://golang.org/pkg/crypto/tls/#Config.GetCertificate
+// and can be used to hot-reload a certificate/key pair.
+//
+// For hot-reloading the Watch method must be started explicitly.
+type CertReloader struct {
+	certPath, keyPath string
+	interval          time.Duration
+
+	mu   sync.RWMutex // protects the fields below
+	cert *tls.Certificate
+	hash uint64
+}
+
+func NewCertReloader(certPath, keyPath string, interval time.Duration) (*CertReloader, error) {
+	r := &CertReloader{
+		certPath: certPath,
+		keyPath:  keyPath,
+		interval: interval,
+	}
+
+	if err := r.reload(); err != nil {
+		return nil, fmt.Errorf("error loading certificates: %v", err)
+	}
+
+	return r, nil
+}
+
+// Watch watches the configured certificate and key path and blocks the current goroutine
+// until the scenario context is done or an error occurred during reloading.
+func (r *CertReloader) Watch(ctx context.Context) error {
+	t := time.NewTicker(r.interval)
+
+	for {
+		select {
+		case <-t.C:
+		case <-ctx.Done():
+			return nil
+		}
+
+		if err := r.reload(); err != nil {
+			return fmt.Errorf("reloading failed: %v", err)
+		}
+	}
+}
+
+func (r *CertReloader) reload() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	var err error
+
+	crt, err := ioutil.ReadFile(r.certPath)
```

Comment:
Unlikely but I think we should check both errors here

Reply:
oh, doh, i wanted to achieve a running err check, silly me, i'll add an err check here too obviously, good catch :+1: